cmd/juju/model/defaultscommand.go

```diff
@@ -97,120 +127,425 @@ func (c *defaultsCommand) SetFlags(f *gnuflag.FlagSet) {
 		}
-		for _, k := range args {
+
+		for _, k := range resetKeys {
 			if k == config.AgentVersionKey {
 				return errors.Errorf("%q cannot be reset", config.AgentVersionKey)
 			}
+			if strings.Contains(k, "=") {
+				return errors.Errorf(
+					`--reset accepts a single key "a" or comma delimited keys "a,b,c", received: %q`, k)
+			}
 		}
-		c.keys = args
-
-		c.action = c.resetDefaults
-		return nil
+		c.resetKeys = resetKeys
 	}
+	return nil
+}
 
-	if len(args) > 0 && strings.Contains(args[0], "=") {
-		// We're setting defaults.
-		options, err := keyvalues.Parse(args, true)
+// parseArgsForRegion parses args to check if the first arg is a region and
+// returns the appropriate remaining args.
+func (c *defaultsCommand) parseArgsForRegion(args []string) ([]string, error) {
+	var err error
+	if len(args) > 0 {
+		// determine if the first arg is cloud/region or region and return
+		// appropriate positional args.
+		args, err = c.parseCloudRegion(args)
 		if err != nil {
-			return errors.Trace(err)
-		}
-		c.values = make(attributes)
-		for k, v := range options {
-			if k == config.AgentVersionKey {
-				return errors.Errorf(`%q must be set via "upgrade-juju"`, config.AgentVersionKey)
-			}
-			c.values[k] = v
+			return nil, errors.Trace(err)
 		}
+	}
+	return args, nil
+}
 
-		c.action = c.setDefaults
-		return nil
+// parseCloudRegion examines args to see if the first arg is a cloud/region or
+// region. If not it returns the full args slice. If it is then it sets cloud
+// and/or region on the object and sends the remaining args back to the caller.
+func (c *defaultsCommand) parseCloudRegion(args []string) ([]string, error) {
+	var cloud, region string
+	cr := args[0]
+	// Must have no more than one slash and it must not be at the beginning or end.
+	if strings.Count(cr, "/") == 1 && !strings.HasPrefix(cr, "/") && !strings.HasSuffix(cr, "/") {
+		elems := strings.Split(cr, "/")
+		cloud, region = elems[0], elems[1]
+	} else {
+		region = cr
+	}
 
+	valid, err := c.validCloudRegion(cloud, region)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if !valid {
+		return args, nil
 	}
-	// We're getting defaults.
-	val, err := cmd.ZeroOrOneArgs(args)
+	return args[1:], nil
+}
+
+// validCloudRegion checks that region is a valid region in cloud, or default cloud
+// if cloud is not specified.
+func (c *defaultsCommand) validCloudRegion(cloudName, region string) (bool, error) {
+	var (
+		isCloudRegion bool
+		cloud         jujucloud.Cloud
+		cTag          names.CloudTag
+		err           error
+	)
+
+	root, err := c.newAPIRoot()
 	if err != nil {
-		return errors.New("can only retrieve a single value, or all values")
+		return false, errors.Trace(err)
 	}
-	if val != "" {
-		c.keys = []string{val}
+	cc := c.newCloudAPI(root)
+	defer cc.Close()
+
+	if cloudName == "" {
+		cTag, err = cc.DefaultCloud()
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+	} else {
+		cTag = names.NewCloudTag(cloudName)
```

Comment:
Thanks I didn't know that. Seems a bad idea to panic though. Shouldn't it 
- normalize and return a tag,
- also return an error,
- or be name MustNewCloudTag?

Added a check and early return if the name is not a valid cloud name.